### PR TITLE
Producer ignorable events

### PR DESCRIPTION
**Pull request by jfragoulis**
From [confluent-kafka-go documentation](https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/kafka.go#L74-L78):

```go
// * Producing is an asynchronous operation so the client notifies the application
// of per-message produce success or failure through something called delivery reports.
// Delivery reports are by default emitted on the `.Events()` channel as `*kafka.Message`
// and you should check `msg.TopicPartition.Error` for `nil` to find out if the message
// was succesfully delivered or not.
```

However, what they do not mention is that the `.Events()` channel does not always consist of `*kafka.Message`s. This is not so obvious, however they somewhat admit it in the [how to setup the producer example](https://github.com/confluentinc/confluent-kafka-go/blob/master/examples/producer_channel_example/producer_channel_example.go#L48-L66):

```go
switch ev := e.(type) {
case *kafka.Message:
  ...
  return

default:
  fmt.Printf("Ignored event: %s\n", ev)
```

Not only do they say "not all objects in the events channel are kafka messages", but they are also ignorrable (you can optionally log them).

What we present as _producer errors_ are actually these ignorable events, and not _delivery reports_ about delivery failures. We do this in the [producer](https://github.com/skroutz/rafka/blob/v0.6.1/producer.go#L102) and this was most likely done because of a [misunderstanding](https://github.com/skroutz/rafka/commit/04ac92c71cac367325ed422be5f5b7b193ec0970) of what these non-kafka-messages were.

The message is:
```
Sep  3 01:07:50 foo rafka[9730]: [producer-49582] 2020/09/03 01:07:50 Error consuming delivery event: Unknown event type (broker:9092/4: Receive failed: Connection reset by peer (after 1199998ms in state UP))
```

The `Error consuming delivery event: Unknown event type` part is from us, from rafka.
The `Receive failed` part is from [librdkafka](https://github.com/edenhill/librdkafka/blob/v0.11.6/src/rdkafka_broker.c#L1432-L1433).
The `Connection reset by peer` part is from tcp, from when [recv](https://github.com/edenhill/librdkafka/blob/master/src/rdkafka_transport.c#L303-L305) returns [ECONNRESET](https://linux.die.net/man/3/recv) . This means that the connection with the broker (peer) has been lost. 
The `(after 1199998ms in state UP)` part is also from librdkafka.

### Q1 Why/how do we get this string in the `.Events()` channel?

`.Events()` consist of many types of structures as we can see from the [poller](https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/producer.go#L606-L621) and the [eventPoll](https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/event.go#L154-L358) method.

### Q2 Why are we getting those disconnections?

The librdkafka devs have already made a [list](https://github.com/edenhill/librdkafka/wiki/FAQ#why-am-i-seeing-receive-failed-disconnected) of things that could be happening. While the first bullet on that list sounds like the most likely scenario, it is worthy of mention, that they have somehow tried to patch up that behavior with [this commit](https://github.com/edenhill/librdkafka/commit/bea2d634459a18d970fea29e69329efaa294101b):

```
The broker will disconnect clients that have not sent any protocol requests
within `connections.max.idle.ms` (broker configuration propertion, defaults
to 10 minutes), but there is no fool proof way for the client to know that it
was a deliberate close by the broker and not an error. To avoid logging these
deliberate idle disconnects as errors the client employs some logic to try to
classify a disconnect as an idle disconnect if no requests have been sent in
the last `socket.timeout.ms` or there are no outstanding, or
queued, requests waiting to be sent. In this case the standard "Disconnect"
error log is silenced (will only be seen with debug enabled).
```

It seems to me that if what we are seeing is indeed idle connection reaping, it should be handled more gracefully.
However, even the developers actually [suggest that we do not take that behavior for granted](https://github.com/edenhill/librdkafka/commit/bea2d634459a18d970fea29e69329efaa294101b#diff-1fe6ecc0e3e1fe5d73d83bc8e8381822c5cc10b0d96306fddab9d46ef671b576R486-R491) in the comments.
 
My suggestion is to look into the options provided and maybe adjust, after having more data regading whether it is the broker that disconnects because of idling etc (for example increase the `socket.timeout.ms`).

**This is most probably something we can ignore and no messages are being lost.**

### CODE REVIEW

**Comment by agis** [comment]
Great write-up! 💯 

IIRC the reason we decided to treat non-delivery events as errors was that we had not much experience with librdkafka back then and we observed that no such events were indeed reporter. So we (wrongly) assumed that this would be an exceptional case that generally shouldn't happen and therefore that we should be notified about it.

Apparently our assumptions were wrong. As you suggested, the situation might be improved with https://github.com/edenhill/librdkafka/commit/bea2d634459a18d970fea29e69329efaa294101b (note that we already set `log.connection.close=false`.

This looks good to me!

---------------------

P.S. The CI failures are due to the changes in `docker-compose.yml`. I know, it works on some systems and breaks in others, depending on the installed Docker version. I'd suggest to leave this unchanged and make any potential improvements in a separate PR.

**Comment by agis** [comment]
On a second thought, I actually think we should _only_ ignore "Connection reset by peer" errors, not all errors.

I just realized that we recently received other delivery errors that we would _not_ like to ignore (i.e. we should be notified of those):

- `No route to host (after 1079ms in state CONNECT))`
- `Connection refused (after 24ms in state CONNECT))`

File: producer.go [hunk 1]
```diff
@@ -69,13 +69,12 @@ func (p *Producer) Produce(msg *rdkafka.Message) error {
...
 func (p *Producer) Flush(timeoutMs int) int {
 	return p.rdProd.Flush(timeoutMs)
-
 }
 
 // Close stops p after flushing any buffered messages. It is a blocking
 // operation.
 func (p *Producer) Close() {
-	unflushed := p.rdProd.Flush(5000)
+	unflushed := p.Flush(5000)
```
Comment (kpelelis) [inline producer.go:77]: How is this change related to the actual issue?

File: producer.go [hunk 2]
```diff
@@ -98,8 +97,7 @@ func (p *Producer) consumeDeliveries() {
 				atomic.AddUint64(&stats.producerErr, 1)
 			}
 		} else {
-			p.log.Printf("Error consuming delivery event: Unknown event type (%s)", ev)
-			atomic.AddUint64(&stats.producerErr, 1)
+			p.log.Printf("Ignored event (%s)", ev)
```
Comment (kpelelis) [inline producer.go:100]: Since there is a [comprehensive list](https://github.com/edenhill/librdkafka/wiki/FAQ#why-am-i-seeing-receive-failed-disconnected) of the things that could have gone wrong, I think we should evaluate which of this are important and which are not.

**Review by kpelelis — CHANGES_REQUESTED** [review]
Kudos on the really detailed writeup :100:. I do not have a lot of experience with the Kafka ecosystem in general, so take my suggestions with a grain of salt. My main concern with this change is the fact that rafka used to check for errors (even in an incorrect way) and now it doesn't. Instead of removing this feature which can introduce false positives onto the monitoring side of things, I suggest that you also [check for errors](https://github.com/confluentinc/confluent-kafka-go/blob/master/examples/producer_channel_example/producer_channel_example.go#L54-L56)

File: producer.go (hunk 1 above)
Comment (jfragoulis) [reply producer.go:77]: It's not

Comment (jfragoulis) [reply producer.go:77]: That's why it is a separate commit, I can just cherry pick to master before merging, for clarity.

**Comment by jfragoulis** [comment]
With https://github.com/skroutz/rafka/pull/90/commits/39e8515971788f47eab1b8bc4ffbe659fc13f1fa I am exposing a way to handle connection errors.

This could also be driven in several directions (for example ignore broken connections with peers) etc.

I suggest, at first we leave this the way it see and observer the error rate and type.

The important thing about this PR is that we will not mark connection errors are delivery errors.

**Comment by jfragoulis** [comment]
> I suggest that you also [check for errors](https://github.com/confluentinc/confluent-kafka-go/blob/master/examples/producer_channel_example/producer_channel_example.go#L54-L56)

This is already being done.

File: producer.go (hunk 1 above)
Comment (kpelelis) [reply producer.go:77]: Yes that is a great idea

File: producer.go [hunk 3]
```diff
@@ -98,8 +97,8 @@ func (p *Producer) consumeDeliveries() {
 				atomic.AddUint64(&stats.producerErr, 1)
 			}
 		} else {
-			p.log.Printf("Error consuming delivery event: Unknown event type (%s)", ev)
-			atomic.AddUint64(&stats.producerErr, 1)
+			p.log.Printf("Connection error (%s)", ev)
```
Comment (agis) [inline producer.go:100]: Let's not assume that these are connection errors (there's no such guarantee anyway). The previous message looked fine to me.

File: producer.go [hunk 4]
```diff
@@ -98,8 +97,8 @@ func (p *Producer) consumeDeliveries() {
...
+			atomic.AddUint64(&stats.connectionErr, 1)
```
Comment (agis) [inline producer.go:101]: We want to ignore "Connection reset by peer" events (we don't want those incrementing the counter).

**Comment by jfragoulis** [comment]
A note on the commits: The final PR will consist of a single commit.
The https://github.com/skroutz/rafka/pull/90/commits/eaabe0630c5b060a6bd221d7a4f88d060c1ae299 will be cherry-picked to master separately since it has nothing to do with the PR, I just wanted to be shown.

**Review by NikosGkotsis — APPROVED** [review]

File: stats.go [hunk 5]
```diff
@@ -30,10 +31,14 @@ func (s *Stats) toRedis() []interface{} {
 		strconv.FormatUint(atomic.LoadUint64(&s.producerUnflushed), 10),
 		"producer.delivery.errors",
 		strconv.FormatUint(atomic.LoadUint64(&s.producerErr), 10),
+		"producer.connection.errors",
```
Comment (oliakremmyda) [inline stats.go:34]: We could probably name this `producer.unknown.events` to be more clear on reports.

**Review by oliakremmyda — APPROVED** [review]
LGTM 👍

**Comment by jfragoulis** [comment]
Τα tests σπανε με librdkafka 0.11.6 και στο μαστερ, δεν ξερω γιατί, δε μπορώ να το βρω σε αυτή τη φάση.

File: producer.go [hunk 6]
```diff
@@ -99,7 +98,13 @@ func (p *Producer) consumeDeliveries() {
 			}
 		} else {
 			p.log.Printf("Error consuming delivery event: Unknown event type (%s)", ev)
-			atomic.AddUint64(&stats.producerErr, 1)
+
+			// Ignore 'connection reset by peer' errors since they manifest because
+			// of the connection reaper.
```
Comment (agis) [inline producer.go:103]: Should we also mention explicitly that these are non-errors? The "since they manifest because of the connection reaper" part doesn't explain much.

File: producer.go [hunk 7]
```diff
@@ -99,7 +98,13 @@ func (p *Producer) consumeDeliveries() {
 			}
 		} else {
 			p.log.Printf("Error consuming delivery event: Unknown event type (%s)", ev)
```
Comment (agis) [inline producer.go:100]: We shouldn't log "connection reset by peer" events.

File: producer.go [hunk 8]
```diff
@@ -99,7 +98,13 @@ func (p *Producer) consumeDeliveries() {
...
-			atomic.AddUint64(&stats.producerErr, 1)
+
+			// Ignore 'connection reset by peer' errors since they manifest because
+			// of the connection reaper.
+			// See https://github.com/edenhill/librdkafka/wiki/FAQ#why-am-i-seeing-receive-failed-disconnected.
+			if !strings.Contains(ev.String(), "Connection reset by peer") {
```
Comment (agis) [inline producer.go:105]: For readability and a more idiomatic version of this, consider the following:

```go
// ...
} else {
  if string.Contains(ev.String(), "Connection reset by peer") {
    continue
  }

  // log
  // increment counter
}
```

File: stats.go [hunk 9]
```diff
@@ -22,6 +22,7 @@ import (
 type Stats struct {
 	producerUnflushed uint64
 	producerErr       uint64
+	unknownEventErr   uint64
```
Comment (agis) [inline stats.go:25]: Since these are not errors per-se and also for consistency with the other metrics, perhaps this should be named instead: `producerUnknownEvents`.

File: stats.go [hunk 10]
```diff
@@ -30,10 +31,13 @@ func (s *Stats) toRedis() []interface{} {
 		strconv.FormatUint(atomic.LoadUint64(&s.producerUnflushed), 10),
 		"producer.delivery.errors",
 		strconv.FormatUint(atomic.LoadUint64(&s.producerErr), 10),
+		"producer.unknown.errors",
```
Comment (agis) [inline stats.go:34]: Ditto: `producer.unknown.events`.

File: producer.go (hunk 8 above)
Comment (agis) [inline producer.go:105]: What's the value returned by `ev.String()`? Maybe we could instead use instead `strings.HasPrefix` to be more explicit and err on the safe side (i.e. reducing the possibility of ignoring something we shouldn't).

Comment (jfragoulis) [reply producer.go:105]: From the logs we can tell that exact value of ev.String() is `broker:9092/4: Receive failed: Connection reset by peer (after 1199998ms in state UP)`.

Comment (agis) [reply producer.go:105]: OK nvm then. Ignore my comment.

Comment (jfragoulis) [reply producer.go:105]: OK for the continue idiom, but I still want to log this.

File: producer.go (hunk 6 above)
Comment (jfragoulis) [reply producer.go:103]: But they are errors, the are connection errors, tcp errors. It is just that they are ignorable by us in this context.

Comment (agis) [reply producer.go:103]: Let me rephrase: should we elaborate more on why we treat them as non-errors (as opposed to merely referring to the connection reaper, which doesn't provide much context)?

**Review by agis — APPROVED** [review]
